.metadata-tools/metadataanalyzercreator.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io/fs"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"strings"
+)
+
+func main() {
+	flag.Parse()
+	root := flag.Arg(0)
+
+	fset := token.NewFileSet()
+	pkgs := make(map[string]string)
+
+	err := filepath.WalkDir(root, func(path string, file fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if file.IsDir() {
+			if file.Name() == "vendor" || file.Name() == "tests" || file.Name() == "internal" {
+				return fs.SkipDir
+			}
+			return nil
+		}
+
+		if filepath.Ext(path) != ".go" {
+			return nil
+		}
+
+		componentType := ""
+		packageName := ""
+		skip := true
+		dir := filepath.Dir(path)
+		for dir != "." && !strings.HasSuffix(dir, "components-contrib") {
+			if !skip {
+				packageName = filepath.Base(dir) + "/" + packageName
+			} else {
+				packageName = filepath.Base(dir)
+			}
+			skip = false
+			dir = filepath.Dir(dir)
+
+			switch filepath.Base(dir) {
+			case "bindings", "pubsub", "state", "secretstores", "middleware", "workflows", "lock", "configuration", "crypto", "nameresolution":
+				componentType = filepath.Base(dir)
+			}
+		}
+
+		parsedFile, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
+		if err != nil {
+			log.Printf("could not parse %s: %v", path, err)
+			return nil
+		}
+
+		var method string
+		var methodFinderErr error
+		methodFound := false
+
+		switch componentType {
+		// Only the component types listed here implement the GetComponentMetadata method today
+		case "secretstores":
+			method, methodFinderErr = getConstructorMethod("secretstores.SecretStore", parsedFile)
+			if methodFinderErr == nil {
+				methodFound = true
+			}
+		case "state":
+			method, methodFinderErr = getConstructorMethod("state.Store", parsedFile)
+			if methodFinderErr == nil {
+				methodFound = true
+			}
+		}
+
+		if methodFound {
+			pkgs[packageName] = method
+		}
+
+		return nil
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// now we have a list of packages that implement the GetComponentMetadata method
+	// now write our go program to check the metadata of each component and compare it againts the yaml metadata
+
+	importBlock := ""
```

Comment:
Although this works, I wonder if putting the code in a template is better. Like, create a separate file, load it from disk, and use the Go template syntax. May be easier to maintain long-term. https://pkg.go.dev/text/template

Reply:
I'll take a look.